src/server.rs

```diff
@@ -207,6 +207,316 @@ impl Server for ServerChain {
+            match parsed {
+                None => Err($crate::RpcError::invalid_params(
+                    Some("Expected parameters".to_owned()))),
+                Some(x) => x,
+            }
+        }
+    };
+
+    // Returns a Result<T, RpcError>
+    ($params:expr) => {
+        {
+            let parsed = $crate::server::FromParams::from_params($params);
+            match parsed {
+                None => Err($crate::RpcError::invalid_params(
+                    Some("Expected parameters".to_owned()))),
+                Some(x) => x,
+            }
+        }
+    }
+}
+
+/// Will exit the function if the parsing fails
+/// expands to a T or Option<T>, will `return` with Option<Result<T, RpcError>> on errors
+// TODO: docs missing
+// TODO: add a tokio-jsonrpc-derive that will do similar code for an already
+#[macro_export]
+macro_rules! parse_params {
+    //
+    // internal
+    //
+
+    // terminal rule
+    (@parse [$params:expr, $(#[$attr:meta])*]
+            $members:tt
+            [$($field_name:ident),*]
+            { $(,)* }
+    ) => {
+        {
+            $(#[$attr])*
+            #[derive(Deserialize)]
+            struct X $members
+
+            fn __parse(params: Option<$crate::Params>) -> Option<Result<X, RpcError>> {
+                match params {
+                    Some($crate::Params::Positional(mut xs)) => {
+                        // Turn the positions into a partial object
+                        // We can leverage any #[serde(default)] by letting serde do the work
+                        let mut object = $crate::macro_exports::Map::new();
+
+                        // Review: do we need to allocate an array
+                        //         or can we just use the macro loop?
+                        let fields = vec![$( stringify!($field_name)),*];
+                        for f in fields {
+                            if ! xs.is_empty() {
+                                object.insert(f.into(), xs.remove(0));
+                            }
+                        }
+
+                        let value = $crate::macro_exports::Value::Object(object);
+                        Some($crate::macro_exports
+                            ::from_value::<X>(value)
+                            .map_err(|err|
+                                $crate::message::RpcError::parse_error(format!("{}", err))))
+                    },
+                    Some($crate::Params::Named(x)) => {
+                        let value = $crate::macro_exports::Value::Object(x);
+                        Some($crate::macro_exports
+                            ::from_value::<X>(value)
+                            .map_err(|err|
+                                $crate::message::RpcError::parse_error(format!("{}", err))))
+                    }
+                    None => {
+                        // Review: should we try to parse an object where all fields are missing?
```

Comment:
That comment is from before the **derive**. It's referring to this code:

```rust
// Try to parse it as if we received no arguments, if we fail DO NOT error
let value = ::tokio_jsonrpc::macro_exports::Value::Null;
match ::tokio_jsonrpc::macro_exports::from_value(value) {
    Ok(x) => Some(Ok(x)),
     Err(_) => None,
}
```

That code will work on the macro too for unit structs but unfortunatly we can't ask if `T: Default` on the macro to pull exactly same trick I have on the derive. 

We could add an option for it, or just tell people to define the struct and use `#[derive(Params)`  for complex use cases.